ip-utils.h

```diff
@@ -0,0 +1,139 @@
+/*! \file    ip-utils.h
+ * \author   Johan Ouwerkerk <jm.ouwerkerk@gmail.com>
+ * \copyright GNU General Public License v3
+ * \brief    IP address related utility functions (headers)
+ * \details  Provides functions to query for network devices with a given device name or address.
+ * \p Devices may be looked up by either a device name or by the IPv4 or IPv6 address of the configured network interface.
+ * This functionality may be used to bind to user configurable network devices instead of relying on unpredictable implementation defined defaults.
+ *
+ * \p Parsing IPv4/IPv6 addresses is robust against malformed input.
+ *
+ * \see man 3 getifaddrs
+ * \see man 3 inet_pton
+ * \ingroup core
+ * \ref core
+ */
+
+#ifndef _JANUS_IP_UTILS_H
+#define _JANUS_IP_UTILS_H
+
+#include <ifaddrs.h>
+#include <netinet/in.h>
+
+/** @name Janus helper methods to match names and addresses with network interfaces/devices.
+ */
+///@{
+typedef enum JanusNetworkQueryOptions {
+	NoMatch = 0, NameOnlyMatch = 1, IPv4OnlyMatch = 2, IPv6OnlyMatch = 4, AnyIPMatch = 6,  AnyMatch = 7
+} JanusNetworkQueryOptions;
+
+/*!
+ * \brief Internal object representation of a network device query (configuration).
+ */
+typedef struct janus_network_query_config {
+	const char * device_name;
+	JanusNetworkQueryOptions mode;
+	struct in_addr ipv4;
+	struct in6_addr ipv6;
+} janus_network_query_config;
+
+/*!
+ * \brief Structure to hold network addresses in a tagged union which should be IPv4 and IPv6 compatible.
+ * Use the `family` member (either `AF_INET` or `AF_INET6`) to determine which type of address is contained.
+ * \see man 7 ip
+ * \see man 7 ipv6
+ * \see `janus_get_network_device_address()`
+ */
+typedef struct janus_network_address {
+	/*!
+	* Should be either`AF_INET` for IPv4 or `AF_INET6` for IPv6.
+	*/
+	int family;
+	union {
+		struct in_addr ipv4;
+		struct in6_addr ipv6;
+	};
+} janus_network_address;
+
+/*!
+ * \brief Initialise a network device query.
+ * \param user_value the user-supplied string which is supposed to describe either the device name or its IP address.
```

Comment:
The description of each parameter should start with a capital letter, as that's the text doxygen automatically adds for each of them.